src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java

```diff
@@ -171,7 +172,10 @@ public boolean execute(Set<InetAddressAndPort> peers, Function<InetAddressAndPor
         {
             // dc -> missing peer host addresses
             Map<String, List<String>> peersDown = acks.getMissingPeers().stream()
-                                                      .collect(groupingBy(peerToDatacenter::get,
+                                                      .collect(groupingBy(peer -> {
+                                                                              String dc = peerToDatacenter.get(peer);
+                                                                              return dc != null ? dc : StringUtils.defaultString(getDatacenterSource.apply(peer), "unknown");
```

Comment:
can be reworked like this:

```
.collect(groupingBy(peer -> peerToDatacenter.getOrDefault(peer,
                                                                          StringUtils.defaultString(getDatacenterSource.apply(peer),
                                                                                                    "unknown")),
                                                                          mapping(InetAddressAndPort::getHostAddressAndPort,
                                                                                  toList())));
```

Reply:
Yeah, I was trying to avoid calling the snitch. I pushed an update. Java 9 has an API to do what I'm trying to do here: https://docs.oracle.com/javase/9/docs/api/java/util/Objects.html#requireNonNullElseGet-T-java.util.function.Supplier-

However, I'm not sure if apache commons or guava offer something similar. I will spend some time looking for that.

Reply:
Ok. I am using `Optional.ofNullable` now. Let me know what you think @smiklosovic

Reply:
we could use only JDK APIs, like this alternatively:

```java
Optional.ofNullable(peerToDatacenter.get(peer)).orElseGet(() -> Optional.ofNullable(getDatacenterSource.apply(peer)).orElse("unknown"))
```

Reply:
I think it has to compile on Java 8 for sure. I recently hit a bug when a developer used `Set.of` which is not in 8 but in 11 (hence 9 maybe?). So please code against Java 8 in general as that is the lowest common denominator.

Reply:
The goal is here very simple, i.e. try to resolve the value from the map cache, if not try to resolve from snitch, if all attempts fail, use the default "unknown". 
There are so many ways to express it in the code and all correctly. Functional, ternary operator, or just the plain `if`
In this particular scenario, I would rate the plain `if` to have the best readability, then the ternary operator. Functional seems a bit overkill.

Reply:
I agree that plain `if` is more readable. I'm going with that approach